streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableTransformValues.java

```diff
@@ -0,0 +1,148 @@
+        this.queryableName = queryableName;
+    }
+
+    @Override
+    public Processor<K, Change<V>> get() {
+        return new KTableTransformValuesProcessor(transformerSupplier.get());
+    }
+
+    @Override
+    public KTableValueGetterSupplier<K, V1> view() {
+        if (queryableName != null) {
+            return new KTableMaterializedValueGetterSupplier<>(queryableName);
+        }
+
+        return new KTableValueGetterSupplier<K, V1>() {
+            final KTableValueGetterSupplier<K, V> parentValueGetterSupplier = parent.valueGetterSupplier();
+
+            public KTableValueGetter<K, V1> get() {
+                return new KTableTransformValuesGetter(
+                        parentValueGetterSupplier.get(),
+                        transformerSupplier.get());
+            }
+
+            @Override
+            public String[] storeNames() {
+                return parentValueGetterSupplier.storeNames();
+            }
+        };
+    }
+
+    @Override
+    public void enableSendingOldValues() {
+        parent.enableSendingOldValues();
+        sendOldValues = true;
+    }
+
+    private static <K, V, V1> V1 computeValue(final K key,
+                                              final V value,
+                                              final ValueTransformerWithKey<? super K, ? super V, ? extends V1> valueTransformer) {
+        if (value == null) {
+            return null;
+        }
+
+        return valueTransformer.transform(key, value);
+    }
+
+    private class KTableTransformValuesProcessor extends AbstractProcessor<K, Change<V>> {
+        private final ValueTransformerWithKey<? super K, ? super V, ? extends V1> valueTransformer;
+        private KeyValueStore<K, V1> store;
+        private TupleForwarder<K, V1> tupleForwarder;
+
+        private KTableTransformValuesProcessor(final ValueTransformerWithKey<? super K, ? super V, ? extends V1> valueTransformer) {
+            this.valueTransformer = Objects.requireNonNull(valueTransformer, "valueTransformer");
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public void init(final ProcessorContext context) {
+            super.init(context);
+
+            valueTransformer.init(new ForwardingDisabledProcessorContext(context));
+
+            if (queryableName != null) {
+                final ForwardingCacheFlushListener<K, V1> flushListener = new ForwardingCacheFlushListener<>(context, sendOldValues);
+                store = (KeyValueStore<K, V1>) context.getStateStore(queryableName);
+                tupleForwarder = new TupleForwarder<>(store, context, flushListener, sendOldValues);
+            }
+        }
+
+        @Override
+        public void process(K key, Change<V> change) {
+            final V1 newValue = computeValue(key, change.newValue, valueTransformer);
+            final V1 oldValue = sendOldValues ? computeValue(key, change.oldValue, valueTransformer) : null;
```

Comment:
Not sure about this... If the operator is stateful, the computation might depend on the state. Thus, this might not compute the correct `oldValue`... 

On the other hand, I am wondering if we need the old value in the first place -- also, the old value would be store in the result KTable state (if there is any) -- thus, maybe using the result KTable state might be the better option (maybe we even need to force a materialization if we need to send `oldValue`.

WDYT?

Reply:
There is plenty of scope for computing the wrong old value here. I've changed the code to make the computation of old value very explicit. 

Materialized tables are easy. Where unmaterialized tables are mixed with stateful transformations it is easy for users to break downstream aggregates. (There's a test that shows this). I've documented this in the Java doc. I *think* it still makes sense to have the ability to pass in additional stateStoreNames for the  non-materialized `transformValues` method call. Not allowing them would remove a big 'gotcha' but may be overly restrictive too.

No doubt the wording of the Java docs around this issue will need finessing, and you may prefer me to remove the 'prove this breaks tests', but just let me know.